src/v/kafka/server/quota_manager.cc

```diff
@@ -254,37 +269,67 @@ ss::future<> quota_manager::record_fetch_tp(
-    auto it = co_await maybe_add_and_retrieve_quota(quota_id, now);
-    it->second.tp_fetch_rate.record(bytes, now);
+    auto fetch_limit = get_client_target_fetch_tp_rate(quota_id);
+    if (!fetch_limit) {
+        co_return;
+    }
+    auto delay [[maybe_unused]] = co_await maybe_add_and_retrieve_quota(
+      quota_id, now, [bytes](quota_manager::client_quota& cq) {
+          if (!cq.tp_fetch_rate.has_value()) {
+              return clock::duration::zero();
+          }
+          auto& fetch_tracker = cq.tp_fetch_rate.value();
+          fetch_tracker.record(bytes);
+          return clock::duration::zero();
+      });
 }
 
-ss::future<throttle_delay> quota_manager::throttle_fetch_tp(
+ss::future<clock::duration> quota_manager::throttle_fetch_tp(
   std::optional<std::string_view> client_id, clock::time_point now) {
     auto quota_id = get_client_quota_id(
       client_id, _target_fetch_tp_rate_per_client_group());
-    auto target_tp_rate = get_client_target_fetch_tp_rate(quota_id);
-
-    if (!target_tp_rate) {
-        co_return throttle_delay{};
+    if (!get_client_target_fetch_tp_rate(quota_id)) {
+        co_return clock::duration::zero();
     }
-    auto it = co_await maybe_add_and_retrieve_quota(quota_id, now);
-    it->second.tp_fetch_rate.maybe_advance_current(now);
-    auto delay_ms = throttle(
-      quota_id, *target_tp_rate, now, it->second.tp_fetch_rate);
-    co_return throttle_delay{.duration = delay_ms};
+
+    auto delay = co_await maybe_add_and_retrieve_quota(
+      quota_id, now, [now](quota_manager::client_quota& cq) {
+          if (!cq.tp_fetch_rate.has_value()) {
+              return clock::duration::zero();
+          }
+          auto& fetch_tracker = cq.tp_fetch_rate.value();
+          return fetch_tracker.update_and_calculate_delay<clock::duration>(now);
+      });
+
+    co_return cap_to_max_delay(quota_id, delay);
 }
 
 // erase inactive tracked quotas. windows are considered inactive if
 // they have not received any updates in ten window's worth of time.
-void quota_manager::gc(clock::duration full_window) {
-    auto now = clock::now();
-    auto expire_age = full_window * 10;
-    // c++20: replace with std::erase_if
-    absl::erase_if(
-      _client_quotas,
-      [now, expire_age](const std::pair<ss::sstring, client_quota>& q) {
-          return (now - q.second.last_seen) > expire_age;
-      });
+void quota_manager::gc() {
+    vassert(
+      ss::this_shard_id() == _client_quotas.shard_id(),
+      "gc should only be performed on the owner shard");
+    auto full_window = _default_num_windows() * _default_window_width();
+    ssx::background
+      = ssx::spawn_with_gate_then(
+          _gate,
+          [full_window, _client_quotas = std::ref(_client_quotas)]() {
+              auto now = clock::now();
+              auto expire_age = full_window * 10;
+              auto new_map = client_quotas_map_t{*_client_quotas.get().local()};
```

Comment:
This copy is unfortunate, but I see it goes away in a future commit.

Reply:
I've reworked that commit now to just comment out the gc code and have gc implemented only once in the later commit.

Reply:
> I've reworked that commit now to just comment out the gc code and have gc implemented only once in the later commit.

That's probably ok, but I guess some more reordering of commits could avoid it.